File: libjs/wController/catalogue.js
"use strict";

var Controller = require("./controller");
var AbstractOrder = require("../wContainer/abstractorder");

var Uint64 = require("../wType/uint64");
var Vocabulary = require("../wType/vocabulary");
var Boolean = require("../wType/boolean");
var String = require("../wType/string");

var IdOrder = AbstractOrder.template(Uint64);

var Catalogue = Controller.inherit({
    "className": "Catalogue",
    "constructor": function(addr, options) {
        Controller.fn.constructor.call(this, addr);
        
        this._hasSorting = false;
        this._hasFilter = false;
        this._filter = undefined;
        this.data = new IdOrder(true);
        
        if (options.sorting) {
            this._hasSorting = true;
            this._sorting = new Vocabulary();
            this._sorting.insert("ascending", new Boolean(options.sorting.ascending));
            this._sorting.insert("field", new String(options.sorting.field));
        }
        if (options.filter) {
            var filter = new Vocabulary();
            for (var key in options.filter) {
                if (options.filter.hasOwnProperty(key)) {
                    filter.insert(key, options.filter[key]);
                }
            }
            if (filter.size()) {
                this._filter = filter;
                this._hasFilter = true;
            } else {
                filter.destructor();
            }
        }
        
        this.addHandler("get");
        this.addHandler("addElement");
        this.addHandler("removeElement");
        this.addHandler("moveElement");
        this.addHandler("clear");
    },
    "destructor": function() {
        this.data.destructor();
        
        if (this._hasSorting) {
            this._sorting.destructor();
        }
        if (this._hasFilter) {
            this._filter.destructor();
        }
        
        Controller.fn.destructor.call(this);
    },
    "addElement": function(element, before) {
        if (before === undefined) {
            this.data.push_back(element);
        } else {
            this.data.insert(element, before);
        }
        
        this.trigger("addElement", element, before);
    },
    "clear": function() {
        this.data.clear();
        this.trigger("clear");
    },
    "_createSubscriptionVC": function() {
        var vc = Controller.fn._createSubscriptionVC.call(this);
        var p = new Vocabulary();
        
        if (this._hasSorting) {
            p.insert("sorting", this._sorting.clone());
        }
        if (this._hasFilter) {
            p.insert("filter", this._filter.clone());
        }
        
        vc.insert("params", p);
        return vc;
    },
    "_h_addElement": function(ev) {
        var data = ev.getData();
        var id = data.at("id").clone();
        var before = undefined;
        if (data.has("before")) {
            before = data.at("before").clone();
        }
        
        this.addElement(id, before);
    },
    "_h_get": function(ev) {
        this.clear();
        
        var data = ev.getData().at("data");
        var size = data.size();
        for (var i = 0; i < size; ++i) {
            this.addElement(data.at(i).clone());
        }
        this.initialized = true;
        this.trigger("data");
    },
    "_h_moveElement": function(ev) {
        var data = ev.getData();
        var id = data.at("id").clone();
        var before = undefined;
        if (data.has("before")) {
            before = data.at("before").clone();
        }
        
        this.data.erase(id);
        if (before === undefined) {
            this.data.push_back(element);
        } else {
            this.data.insert(element, before);
        }
        
        this.trigger("moveElement", id, before);
    },
    "_h_removeElement": function(ev) {
        var data = ev.getData();
        
        this.removeElement(data.at("id").clone());
    },
    "_h_clear": function(ev) {
        this.clear();
    },
    "removeElement": function(id) {
        this.data.erase(id);
        this.trigger("removeElement", id);
    }
});

module.exports = Catalogue;

File: lib/wController/catalogue.cpp
#include "catalogue.h"

uint64_t C::Catalogue::counter = 0;

C::Catalogue::Catalogue(const W::Address p_address, QObject* parent):
    C::Controller(p_address, W::Address({W::String(u"catalogue") += counter++}), parent),
    order(),
    hasSorting(false),
    hasFilter(false),
    hasData(true),
    sorting(0),
    filter(0)
{
    W::Handler* get = W::Handler::create(address + W::Address({u"get"}), this, &C::Catalogue::_h_get);
    W::Handler* addElement = W::Handler::create(address + W::Address({u"addElement"}), this, &C::Catalogue::_h_addElement);
    W::Handler* removeElement = W::Handler::create(address + W::Address({u"removeElement"}), this, &C::Catalogue::_h_removeElement);
    W::Handler* moveElement = W::Handler::create(address + W::Address({u"moveElement"}), this, &C::Catalogue::_h_moveElement);
    W::Handler* clear = W::Handler::create(address + W::Address({u"clear"}), this, &C::Catalogue::_h_clear);
    
    addHandler(get);
    addHandler(addElement);
    addHandler(removeElement);
    addHandler(moveElement);
    addHandler(clear);
}

C::Catalogue::~Catalogue()
{
    if (hasFilter) {
        delete filter;
    }
    
    if (hasSorting) {
        delete sorting;
    }
}

void C::Catalogue::setSorting(const W::String& field, bool ascending)
{
    if (!hasSorting) {
        sorting = new W::Vocabulary();
        hasSorting = true;
    }
    sorting->insert(u"field", field);
    sorting->insert(u"ascending", W::Boolean(ascending));
    
    if (hasData) {
        clearCatalogue();
    }
    
    if (subscribed) {
        getData();
    }
}

void C::Catalogue::clearSorting()
{
    if (hasSorting) {
        delete sorting;
        hasSorting = false;
        
        if (hasData) {
            clearCatalogue();
        }
        
        if (subscribed) {
            getData();
        }
    }
}

void C::Catalogue::addElement(const W::Uint64& id, const W::Uint64& before)
{
    if (before == 0) {
        order.push_back(id);
    } else {
        W::Order<uint64_t>::const_iterator pos = order.find(before);
        order.insert(pos, id);
    }
    
    emit addedElement(id, before);
}

void C::Catalogue::h_get(const W::Event& ev)
{
    if (hasData) {
        clearCatalogue();
    }
    
    const W::Vocabulary& vc = static_cast<const W::Vocabulary&>(ev.getData());
    const W::Vector& ord = static_cast<const W::Vector&>(vc.at(u"data"));
    
    W::Vector::size_type size = ord.size();
    for (uint64_t i = 0; i < size; ++i) {
        const W::Uint64& id = static_cast<const W::Uint64&>(ord.at(i));
        addElement(id);
    }
    hasData = true;
    
    emit data();
}

void C::Catalogue::h_addElement(const W::Event& ev)
{
    const W::Vocabulary& vc = static_cast<const W::Vocabulary&>(ev.getData());
    const W::Uint64& id = static_cast<const W::Uint64&>(vc.at(u"id"));
    if (vc.has(u"before")) {
        const W::Uint64& before = static_cast<const W::Uint64&>(vc.at(u"before"));
        
        addElement(id, before);
    } else {
        addElement(id);
    }
}

void C::Catalogue::clearCatalogue()
{
    order.clear();
    hasData = false;
    emit clear();
}


void C::Catalogue::h_clear(const W::Event& ev)
{
    clearCatalogue();
}

void C::Catalogue::h_removeElement(const W::Event& ev)
{
    const W::Vocabulary& vc = static_cast<const W::Vocabulary&>(ev.getData());
    const W::Uint64& id = static_cast<const W::Uint64&>(vc.at(u"id"));
    
    removeElement(id);
}

void C::Catalogue::removeElement(const W::Uint64& id)
{
    W::Order<uint64_t>::const_iterator pos = order.find(id);
    if (pos == order.end()) {
        emit serviceMessage(QString("Recieved event to remove element with id ") + id.toString().c_str() + " but element under such id isn't present in catalogue, skipping");
        return;
    }
    order.erase(id);
    
    uint64_t pid;
    emit removedElement(pid);
}

W::Vocabulary * C::Catalogue::createSubscriptionVC() const
{
    W::Vocabulary* vc = C::Controller::createSubscriptionVC();
    
    if (hasSorting) {
        vc->insert(u"sorting", sorting->copy());
    }
    
    if (hasFilter) {
        vc->insert(u"filter", filter->copy());
    }
    
    return vc;
}

void C::Catalogue::h_moveElement(const W::Event& ev)
{
    const W::Vocabulary& vc = static_cast<const W::Vocabulary&>(ev.getData());
    const W::Uint64& id = static_cast<const W::Uint64&>(vc.at(u"id"));
    
    W::Order<uint64_t>::const_iterator pos = order.find(id);
    if (pos == order.end()) {
        emit serviceMessage(QString("Recieved event to move element with id ") + id.toString().c_str() + " but element under such id isn't present in catalogue, skipping");
        return;
    }
    
    order.erase(id);
    if (vc.has(u"before")) {
        const W::Uint64& before = static_cast<const W::Uint64&>(vc.at(u"before"));
        
        W::Order<uint64_t>::const_iterator beforePosition = order.find(before);
        if (beforePosition == order.end()) {
            emit serviceMessage(QString("Recieved event to move element with id ") +
                                        id.toString().c_str() + " before element with id " + before.toString().c_str() +
                                        " but element under id " + before.toString().c_str() +
                                        " isn't present in catalogue, inserting to the end");
            
            order.push_back(id);
            emit movedElement(id);
            
            return;
        }
        order.insert(beforePosition, id);
        emit movedElement(id, before);
        
    } else {
        order.push_back(id);
        emit movedElement(id);
    }
}

void C::Catalogue::getData()
{
    W::Vocabulary* vc =  new W::Vocabulary();
    vc->insert(u"params", createSubscriptionVC());
    send(vc, W::Address{u"get"});
}


void C::Catalogue::addRemoteElement(const W::Vocabulary& element) const
{
    send(static_cast<W::Vocabulary*>(element.copy()), W::Address{u"add"});
}

void C::Catalogue::updateRemoteElement(const W::Uint64& id, const W::Vocabulary& newValue) const
{
    W::Vocabulary* vc =  new W::Vocabulary();
    vc->insert(u"id", id);
    vc->insert(u"value", newValue);
    send(vc, W::Address{u"update"});
}
